pipeline/metadata/flatten.py

```diff
@@ -2,210 +2,30 @@
-  # individual values, or for specific combinations of values.
-  flat_headers = []
-  for key, values in headers.items():
-    for value in values:
-      flat_headers.append(key + ': ' + value)
-  return flat_headers
-
-
-def source_from_filename(filepath: str) -> str:
-  """Get the source string from a scan filename.
-
-  Source represents the .tar.gz container which held this file.
-
-  Args:
-    filepath:
-    'gs://firehook-scans/echo/CP_Quack-echo-2020-08-23-06-01-02/results.json'
-
-  Returns:
-    Just the 'CP_Quack-echo-2020-08-23-06-01-02' source
-  """
-  path = os.path.split(filepath)[0]
-  path_end = os.path.split(path)[1]
-  return path_end
-
-
-def format_timestamp(timestamp: str) -> str:
-  """Format the timestamp in ISO8601 format for BigQuery.
-
-  Args:
-    timestamp: Satellite format timestamp
-      "2021-04-18 14:49:01.62448452 -0400 EDT m=+10140.555964129"
-
-  Returns:
-    ISO8601 formatted string "2021-04-18T14:49:01.62448452-04:00"
-  """
-  elements = timestamp.split()
-  date = elements[0]
-  time = elements[1]
-  timezone_hour = elements[2][0:-2]
-  timezone_seconds = elements[2][-2:]
-  return f'{date}T{time}{timezone_hour}:{timezone_seconds}'
-
-
-def _extract_domain_from_sent_field(sent: str) -> Optional[str]:
-  """Get the url out of a 'sent' field in a measurement.
-
-  Args:
-    sent: string like either
-
-      "" meaning the sent packet wasn't recorded.
-      "GET / HTTP/1.1\r\nHost: example5718349450314.com\r\n" (echo/discard)
-      "GET www.bbc.co.uk HTTP/1.1\r\nHost: /content.html\r\n" (discard error)
-      or just "www.apple.com" (HTTP/S)
-
-    Returns: just the url or None
-  """
-  if sent == '':
-    return None
-
-  match = re.search(SENT_PATTERN, sent)
-  if match:
-    path = match.group(1)
-    domain = match.group(2)
-
-    # This is a bug where the domain and path were reversed in content sent.
-    # We do our best to reconstruct the intended url
-    # by swapping them to their intended position
-    # TODO should we do something else instead because the test is invalid?
-    if domain[0] == '/':
-      domain, path = path, domain
-
-    if path == '/':
-      return domain
-    return domain + path
-
-  if ' ' not in sent:
-    return sent
-
-  raise Exception(f"unknown sent field format: {sent}")
-
-
-def _is_control_url(url: Optional[str]) -> bool:
-  return url in CONTROL_URLS
-
-
-def _reconstruct_http_response(row: Row) -> str:
-  """Rebuild the HTTP response as a string from its pieces
-
-    Args:
-      row: a row with the received_status/body/headers fields
-
-    Returns: a string imitating the original http response
-    """
-  full_response = row['received_status'] + '\r\n'
-  for header in row['received_headers']:
-    full_response += header + '\r\n'
-  full_response += '\r\n' + row['received_body']
-  return full_response
-
-
-class FlattenMeasurement(beam.DoFn):
+class FlattenMeasurement(FlattenSatelliteMixin, FlattenHyperquackMixin):
```

Comment:
This DoFn is actually doing too much. It violates the single responsibility principle.
Consider creating separate functions:
- One that just does the JSON parsing and logging.
- One that dispatches to separate PCollections based on the source. Perhaps you can change the file listing logic.

This way you can have two separate FlattenSattelite and FlattenHyperquack that operate on the right files and can inherit directly from DoFn.

Reply:
Example of multiple outputs: https://stackoverflow.com/questions/52075600/python-apache-beam-multiple-outputs-processing

Reply:
So the structure would be something more like

FlattenMeasurements:
- sets up object for DoFn:
- - self.blockpage_matcher
- - self.category_matcher
- has members:
- - parse_satellite_factory
- - parse_hyperquack_factory

and both of those factories take the matchers as constructor inputs? And also contain a parse_base_factory-type thing that also takes the matchers and contains the shared functionality?

Reply:
If we are talking about the approach of having a single `FlattenMeasurements` Transform, I imagine a setup function like:
```py
def setup(self) -> None:
    blockpage_matcher = BlockpageMatcher()
    category_matcher = DomainCategoryMatcher()
    self.hyperquack_flattener = HyperQuackFlattener(blockpage_matcher, category_matcher)
    self.satellite_flattener = SatelliteFlattener(blockpage_matcher, category_matcher)
```

Though you may be able to keep the category logic in the shared logic and not have to pass the matcher to those classes.

BTW, consider flattening and adding annotations as separate pipeline steps. It will should allow Beam to split the work better and you will get clearer performance numbers. But it doesn't have to be in this PR.

Reply:
Switched to using composition instead of inheritance, using pretty much exactly the setup you wrote above.

This is definitely more readable, and now FlattenMeasurement is the only class that knows about DoFn and handles the setup/process logic which is much nicer.